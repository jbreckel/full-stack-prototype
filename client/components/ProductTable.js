import React from 'react'

import UpdateProduct from '../containers/modals/UpdateProduct'
import DeleteProduct from '../containers/modals/DeleteProduct'

const ProductTable = ({ categories, products }) => (
  <table>
    <thead>
      <tr>
        <th>Category</th>
        <th>ID</th>
        <th>Name</th>
        <th>Purchase price</th>
        <th>Sales price</th>
        <th>Edit</th>
        <th>Delete</th>
      </tr>
    </thead>
    <tbody>
      {
        products.map((product) => {
          const { id, name, purchasePrice, salePrice, categoryId } = product
          const category = categories.find(({ id: cid }) => cid === categoryId)
            || { name: 'empty' }
          return (
            <tr key={ id }>
              <td>{ category.name }</td>
              <td>{ id.substring(0, 5) }</td>
              <td>{ name }</td>
              <td>{ purchasePrice }</td>
              <td>{ salePrice }</td>
              <td><UpdateProduct product={ product }>{`Edit ${name}`}</UpdateProduct></td>
              <td><DeleteProduct product={ product }>{`Delete ${name}`}</DeleteProduct></td>
            </tr>
          )
        }
        )
      }
    </tbody>
  </table>
)

export default ProductTable
