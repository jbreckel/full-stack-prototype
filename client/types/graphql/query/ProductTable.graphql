#import '../fragments/Product.graphql'

query ProductTable {
  products {
    ...Product
  }
}
