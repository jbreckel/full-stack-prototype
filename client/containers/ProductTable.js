import { compose } from 'recompose'
import { connect } from 'react-redux'
import { graphql } from 'react-apollo'

import ProductTableQuery from '../types/graphql/query/ProductTable.graphql'
import updateProduct from '../types/graphql/mutation/updateProduct.graphql'
import removeProduct from '../types/graphql/mutation/removeProduct.graphql'

import ProductTable from '../components/ProductTable'

export default compose(
  graphql(updateProduct, {
    props: ({ mutate }) => ({
      updateProduct: (product) => mutate({
        variables: { product },
      }),
    }),
  }),
  graphql(removeProduct, {
    props: ({ mutate }) => ({
      removeProduct: (product) => mutate({
        variables: { product },
      }),
    }),
  }),
  connect(
    ({ categoryView }) => ({
      categories: categoryView.categories,
    }),
  ),
  graphql(ProductTableQuery, {
    options: ({ categories }) => ({
      variables: {
        categories,
      },
    }),
    props: ({ data: { products = [], ...querySubscriptions } }) => ({
      products,
      querySubscriptions,
    }),
  }),
)(ProductTable)
