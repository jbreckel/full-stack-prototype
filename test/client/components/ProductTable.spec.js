import React from 'react'
import { shallow } from 'enzyme'

import ProductTable from '../../../client/components/ProductTable'

jest.mock('../../../client/containers/modals/UpdateProduct', () => 'UpdateProduct')
jest.mock('../../../client/containers/modals/DeleteProduct', () => 'DeleteProduct')

describe('ProductTable', () => {
  it('render', () => {
    const wrapper = shallow(
      <ProductTable
        products={[
          {
            id: 'id-1',
            name: 'name',
            purchasePrice: 14,
            salePrice: 15,
            category: {
              id: 'id',
              name: 'category',
            },
          },
          {
            id: 'id-2',
            name: 'name',
            purchasePrice: 14,
            salePrice: 15,
            category: null,
          },
        ]}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
