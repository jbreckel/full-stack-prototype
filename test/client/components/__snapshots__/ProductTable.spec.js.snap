// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTable render 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Category
      </th>
      <th>
        ID
      </th>
      <th>
        Name
      </th>
      <th>
        Purchase price
      </th>
      <th>
        Sales price
      </th>
      <th>
        Edit
      </th>
      <th>
        Delete
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        category
      </td>
      <td>
        id-1
      </td>
      <td>
        name
      </td>
      <td>
        14
      </td>
      <td>
        15
      </td>
      <td>
        <UpdateProduct
          product={
            Object {
              "category": Object {
                "id": "id",
                "name": "category",
              },
              "id": "id-1",
              "name": "name",
              "purchasePrice": 14,
              "salePrice": 15,
            }
          }
        >
          Edit name
        </UpdateProduct>
      </td>
      <td>
        <DeleteProduct
          product={
            Object {
              "category": Object {
                "id": "id",
                "name": "category",
              },
              "id": "id-1",
              "name": "name",
              "purchasePrice": 14,
              "salePrice": 15,
            }
          }
        >
          Delete name
        </DeleteProduct>
      </td>
    </tr>
    <tr>
      <td>
        empty
      </td>
      <td>
        id-2
      </td>
      <td>
        name
      </td>
      <td>
        14
      </td>
      <td>
        15
      </td>
      <td>
        <UpdateProduct
          product={
            Object {
              "category": null,
              "id": "id-2",
              "name": "name",
              "purchasePrice": 14,
              "salePrice": 15,
            }
          }
        >
          Edit name
        </UpdateProduct>
      </td>
      <td>
        <DeleteProduct
          product={
            Object {
              "category": null,
              "id": "id-2",
              "name": "name",
              "purchasePrice": 14,
              "salePrice": 15,
            }
          }
        >
          Delete name
        </DeleteProduct>
      </td>
    </tr>
  </tbody>
</table>
`;
